### PR TITLE
BHBC-2152: Implement Artifact call to BioHub Artifact Submission API

### PR DESCRIPTION
**Pull request by curtisupshall**
# Overview
Implement SIMS interaction with BioHub Artifact Submission API as described in the workflow here:

 - [SIMS to BioHub Workflows](https://apps.nrs.gov.bc.ca/int/confluence/x/Zpu6C)

   - See diagram titled: “File Artifact Submissions”
## Links to Jira tickets

 - https://quartech.atlassian.net/browse/BHBC-2152

## Description of relevant changes
- Collector system can call the BioHub Artifact Submission API endpoint, sending it one ore more artifacts (files).
- Creates the service(s) (and related repo/util code) needed to facilitate the workflow
- Removes authenticated read from S3 upload util functions
## PR Checklist

A list of items that are good to consider when making any changes.

_Note: this list is not exhaustive, and not all items are always applicable._

### Code

- [ ] New files/classes/functions have appropriately descriptive names and comment blocks to describe their use/behaviour
- [ ] I have avoided duplicating code when possible, moving re-usable pieces into functions
- [ ] I have avoided hard-coding values where possible and moved any re-usable constants to a constants file
- [ ] My code is as flat as possible (avoids deeply nested if/else blocks, promise chains, etc)
- [ ] My code changes account for null/undefined values and handle errors appropriately
- [ ] My code uses types/interfaces to help describe values/parameters/etc, help ensure type safety, and improve readability

### Style

- [ ] My code follows the established style conventions
- [ ] My code uses native material-ui components/icons/conventions when possible

### Documentation

- [ ] I have commented my code sufficiently, such that an unfamiliar developer could understand my code
- [ ] I have added/updated README's and related documentation, as needed

### Tests

- [ ] I have added/updated unit tests for any code I've added/updated
- [ ] I have added/updated the Postman requests/tests to account for any API endpoints I've added/updated

### Linting/Formatting

- [ ] I have run the linter and fixed any issues, as needed  
       _See the `lint` commands in package.json_
- [ ] I have run the formatter and fixed any issues, as needed  
       _See the `format` commands in package.json_

### SonarCloud

- [ ] I have addressed all SonarCloud Bugs, Vulnerabilities, Security Hotspots, and Code Smells

### CODE REVIEW

**Comment by AlfredRosenthal** [comment]
the workflow mentions updating a publish table once the artifacts are uploaded to biohub, [BHBC-2169](https://github.com/bcgov/biohubbc/pull/953) has a new repo/ service for doing that. I think I covered all the publish tables but just a heads up so you don't duplicate the work

**Comment by curtisupshall** [comment]
> the workflow mentions updating a publish table once the artifacts are uploaded to biohub, [BHBC-2169](https://github.com/bcgov/biohubbc/pull/953) has a new repo/ service for doing that. I think I covered all the publish tables but just a heads up so you don't duplicate the work

Good catch! I'll incorporate those changes when your branch gets merged

File: api/src/services/platform-service.ts
Comment (NickPhura) [inline api/src/services/platform-service.ts:84]: Generally our services always require a connection because they are always making database calls, but this service has a number of functions that don't care about the database. 

So we may need to push this connection and related service down into the function itself, so that you can still use this service to get a token, for example, without having to provide a connection that isn't used.

Comment (NickPhura) [reply api/src/services/platform-service.ts:84]: Oh wait, I am mixing this service up with the keycloak service! Ignore this comment haha

File: env_config/env.docker
Comment (NickPhura) [inline env_config/env.docker:78]: Could even include the URL to the CSS site if you want

File: api/.pipeline/config.js
Comment (NickPhura) [inline api/.pipeline/config.js]: Would you mind also adjusting the submission intake path while you're at it. I believe it is now `/queue` rather than `/intake`

File: api/src/services/platform-service.ts
Comment (NickPhura) [inline api/src/services/platform-service.ts]: You could consider making a new function that gets attachments based on an array of ids (select * where in) type call. Then you wouldn't have to filter them here.

**Comment by NickPhura** [comment]
@curtisupshall Gave the code a look over, looks good overall. ~~I think the only thing missing is populating the attachment/report publish tables after the call to biohub is made successfully for each attachment. Probably some code there that overlaps with Alfreds PR.~~

Added a couple of last minute comments, but looks good otherwise!

File: api/src/services/platform-service.test.ts
Comment (KjartanE) [inline api/src/services/platform-service.test.ts]: Wrap up tests files and looks good to go

File: api/.pipeline/config.js
Comment (NickPhura) [inline api/.pipeline/config.js]: Could you update this submission path to be `.../queue` as well, and the one below on line 115.

File: api/src/services/platform-service.ts
Comment (NickPhura) [inline api/src/services/platform-service.ts]: Could also write this as:
```
return Promise.all([...attachmentArtifactPromises, ...reportArtifactPromises]);
```
Should yield the same result

File: api/src/services/attachment-service.ts
Comment (NickPhura) [inline api/src/services/attachment-service.ts]: We should probably remove this before merging. Good idea on how to test it, and could be used locally to demo it is working, but lets not merge it to dev. We'll be starting on the UIs to trigger this next sprint, so it won't be longer before we can run it for real.

File: api/src/services/platform-service.ts
Comment (curtisupshall) [reply api/src/services/platform-service.ts]: When I write `Promise.all([...attachmentArtifactPromises, ...reportArtifactPromises])`, the compiler complains... Admitedly it looks a bit messy, so I'll refactor

File: api/src/repositories/attachment-repository.ts
Comment (NickPhura) [inline api/src/repositories/attachment-repository.ts]: This check could be removed because rows will always be an array. That being said, it is in a bunch of other places from past code, so we could opt to clean them all up separately later, as tech debt. Up to you though.

File: api/src/services/survey-service.test.ts
Comment (NickPhura) [inline api/src/services/survey-service.test.ts]: Can this be removed?

File: api/src/services/platform-service.ts
Comment (NickPhura) [inline api/src/services/platform-service.ts]: Could do the same to the return here as below

File: api/src/repositories/attachment-repository.ts
Comment (NickPhura) [inline api/src/repositories/attachment-repository.ts:307]: Just curious: looks like we only do this for project/survey reports, but not attachments. Any reason? Or just do work with existing interfaces that were designed differently, or something?

**Review by KjartanE — APPROVED** [review]
🥇

**Pull request merged**
